### PR TITLE
feat: dynamicfees keeper as denom resolver

### PR DESCRIPTION
**Pull request by swelf19**
related prs:
https://github.com/neutron-org/neutron-integration-tests/pull/300
https://github.com/neutron-org/neutronjsplus/pull/38

### CODE REVIEW

**Review by pr0n00gler — APPROVED** [review]

**Comment by joldie777** [comment]
Please resolve merge conflicts and attach a successful test run

File: app/upgrades/v4.0.0/upgrades.go [hunk 1]
```diff
@@ -72,6 +93,48 @@ func setMarketMapParams(ctx sdk.Context, marketmapKeeper *marketmapkeeper.Keeper
...
+		MaxBlockUtilization:    0,
+		Window:                 0,
+		FeeDenom:               "",
+		Enabled:                false,
+		DistributeFees:         false,
+	}
+	feemarketState := feemarkettypes.NewState(feemarketParams.Window, feemarketParams.MinBaseGasPrice, feemarketParams.MinLearningRate)
+	err := feemarketKeeper.SetParams(ctx, feemarketParams)
+	if err != nil {
+		return errors.Wrap(err, "failed to to set feemarket params")
```
Comment (joldie777) [inline app/upgrades/v4.0.0/upgrades.go]: Redundant "to"

File: app/upgrades/v4.0.0/upgrades.go [hunk 2]
```diff
@@ -72,6 +93,48 @@ func setMarketMapParams(ctx sdk.Context, marketmapKeeper *marketmapkeeper.Keeper
...
+	}
+	err = feemarketKeeper.SetState(ctx, feemarketState)
+	if err != nil {
+		return errors.Wrap(err, "failed to to set feemarket state")
```
Comment (joldie777) [inline app/upgrades/v4.0.0/upgrades.go]: Redundant "to"

File: app/upgrades/v4.0.0/upgrades.go [hunk 3]
```diff
@@ -72,6 +93,48 @@ func setMarketMapParams(ctx sdk.Context, marketmapKeeper *marketmapkeeper.Keeper
...
 
+// NtrnPrices describes prices of any token in NTRN for dynamic fee resolver
+// TODO: determine actual prices
+var NtrnPrices = sdk.NewDecCoins(sdk.NewDecCoin(params.DefaultDenom, math.OneInt().Mul(math.NewInt(100))))
+
+func setDynamicFeesParams(ctx sdk.Context, dfKeeper *dynamicfeeskeeper.Keeper) error {
+	dfParams := dynamicfeestypes.Params{
+		NtrnPrices: NtrnPrices,
+	}
+	return dfKeeper.SetParams(ctx, dfParams)
```
Comment (joldie777) [inline app/upgrades/v4.0.0/upgrades.go]: Why not
```
if err != nil {
	return errors.Wrap(err, "failed to set dynamic fees params")
}

return nil
```
Just to be consistent with the errors in the `setFeeMarketParams` func below

File: proto/neutron/dynamicfees/v1/genesis.proto [hunk 4]
```diff
@@ -0,0 +1,13 @@
+syntax = "proto3";
+package neutron.dynamicfees.v1;
+
+import "cosmos/base/v1beta1/coin.proto";
```
Comment (joldie777) [inline proto/neutron/dynamicfees/v1/genesis.proto]: Unused import

File: x/dynamicfees/types/tx.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package types
+
+func (q *MsgUpdateParams) ValidateBasic() error {
```
Comment (joldie777) [inline x/dynamicfees/types/tx.go]: I guess it should be named `Validate()`

File: proto/neutron/dynamicfees/v1/genesis.proto (hunk 4 above)
Comment (swelf19) [reply proto/neutron/dynamicfees/v1/genesis.proto]: .

**Review by joldie777 — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Pull request merged**